evalml/tests/conftest.py

```diff
@@ -1016,3 +1023,170 @@ def _imbalanced_data_X_y(problem_type, categorical_columns, size):
+        return {
+            ProblemTypes.REGRESSION: "evalml.pipelines.RegressionPipeline",
+            ProblemTypes.BINARY: "evalml.pipelines.BinaryClassificationPipeline",
+            ProblemTypes.MULTICLASS: "evalml.pipelines.MulticlassClassificationPipeline",
+            ProblemTypes.TIME_SERIES_REGRESSION: "evalml.pipelines.TimeSeriesRegressionPipeline",
+            ProblemTypes.TIME_SERIES_MULTICLASS: "evalml.pipelines.TimeSeriesMulticlassClassificationPipeline",
+            ProblemTypes.TIME_SERIES_BINARY: "evalml.pipelines.TimeSeriesBinaryClassificationPipeline",
+        }[self.problem_type]
+
+    def _patch_method(self, method, side_effect, return_value, pipeline_class_str=None):
+        kwargs = {}
+        if pipeline_class_str is None:
+            pipeline_class_str = self._pipeline_class
+        if side_effect is not None:
+            kwargs = {"side_effect": side_effect}
+        elif return_value is not None:
+            kwargs = {"return_value": return_value}
+        return patch(pipeline_class_str + "." + method, **kwargs)
+
+    @contextlib.contextmanager
+    def test_context(
+        self,
+        score_return_value=None,
+        mock_score_side_effect=None,
+        mock_fit_side_effect=None,
+        mock_fit_return_value=None,
+        predict_proba_return_value=None,
+        optimize_threshold_return_value=0.2,
+    ):
+        """A context manager for creating an environment that patches time-consuming pipeline methods.
+        Sets the mock_fit, mock_score, mock_encode_targets, mock_predict_proba, mock_optimize_threshold attributes.
+
+        Arguments:
+            score_return_value: Passed as the return_value argument of the pipeline.score patch.
+            mock_score_side_effect: Passed as the side_effect argument of the pipeline.score patch. Takes precedence over
+                score_return_value.
+            mock_fit_side_effect: Passed as the side_effect argument of the pipeline.fit patch. Takes precedence over mock_fit_return_value.
+            mock_fit_return_value: Passed as the return_value argument of the pipeline.fit patch.
+            predict_proba_return_value: Passed as the return_value argument of the pipeline.predict_proba patch.
+            optimize_threshold_return_value: Passed as the return value of BinaryClassificationObjective.optimize_threshold patch.
+        """
+        mock_fit = self._patch_method(
+            "fit", side_effect=mock_fit_side_effect, return_value=mock_fit_return_value
+        )
+        mock_score = self._patch_method(
+            "score", side_effect=mock_score_side_effect, return_value=score_return_value
+        )
+
+        # For simplicity, we will always mock predict_proba and _encode_targets even if the problem is not a binary
+        # problem. For time series problems, we will mock these methods in the time series class (self._pipeline_class)
+        # and for non-time-series problems we will use BinaryClassificationPipeline
```

Comment:
Good point. We don't **_have_** to because the only time we call these methods outside of `score` is for threshold tuning. Since mutliclass pipelines don't support threshold tuning and we mock `score` for the `_pipeline_class`, everything works. But the implementation is nicer if we patch the multiclass methods for multiclass pipelines.